src/Chapter3.hs

```diff
@@ -1005,8 +1111,85 @@ properties using typeclasses, but they are different data types in the end.
 Implement data types and typeclasses, describing such a battle between two
 contestants, and write a function that decides the outcome of a fight!
 -}
-
-
+data Knight'Action = KAttack
+                   | HealthPotion
+                   | IncreaseArmor deriving (Show)
+data Monster'Action = MAttack
+                    | RunAway deriving (Show)
+
+-- data Action = KAction Knight'Action | MAction Monster'Action deriving (Show)
+data Knight' = Knight' { knight'Health  :: Health
+                       , knight'Attack  :: Attack
+                       , knight'Armor   :: Armor
+                       , knight'Actions :: [Knight'Action]
+                       } deriving (Show)
+data Monster' = Monster' { monster'Health  :: Health
+                         , monster'Attack  :: Attack
+                         , monster'Actions :: [Monster'Action]
+                         } deriving (Show)
+
+class (Show a) => Fighter a where
+  performAction :: (Fighter b) => a -> b -> (a, b)
+  receiveDamage :: a -> Attack -> a
+  getHealth :: a -> Int
+
+instance Fighter Knight' where
+  performAction :: (Fighter enemy) => Knight' -> enemy -> (Knight', enemy)
+  performAction knight@(Knight' { knight'Health = (Health kHealth)
+                                , knight'Attack = kAttack
+                                , knight'Armor  = (Armor  kArmor)
+                                , knight'Actions = (currentAction:actions)
```

Comment:
This pattern matching is not exhaustive. Even though we know that we are going to use cycle on the actions, it is not reflected in types, so potentially there could be `[]` in here, which could lead to the runtime error 🙂 

This could be solved in several ways:
1. Explicit pattern matching on all cases
2. Using more relevant data type, which should always have at least one elem (see `NonEmpty` standard type)

Reply:
thanks for advice wtih NonEmpty!